server/src/main/java/org/opensearch/action/support/replication/TransportWriteAction.java

```diff
@@ -244,6 +250,11 @@ protected abstract void dispatchedShardOperationOnPrimary(
      */
     @Override
     protected void shardOperationOnReplica(ReplicaRequest request, IndexShard replica, ActionListener<ReplicaResult> listener) {
+        // The call should finish here for cases where the replication call is NoOp
+        if (isNoOp(request, replica)) {
+            ActionListener.completeWith(listener, ReplicaResult::new);
+            return;
+        }
         threadPool.executor(executorFunction.apply(replica)).execute(new ActionRunnable<ReplicaResult>(listener) {
             @Override
             protected void doRun() {
```

Comment:
This still isn't No-op replication IMO :( as checkpoints on replica might still be modified 
The `shardOperationOnReplica` is invoked after acquiring replica permits which ends up performing further mutating actions(updating checkpoints) before finally transferring back the control to this method

Please look at
https://github.com/opensearch-project/OpenSearch/blob/f65e02d1b910bd0a1990868bfa5d12ba829bbbd5/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L3513-L3539

Reply:
Yes, to decouple updating the replication tracker would be a huge effort. I would prefer to handle that incrementally in follow up commits.